`pip install -e .` fails

Running `pip install -e .` fails with:

```
➜ pip install -e .        
Obtaining file:///Users/yuvipanda/code/jupyter-keepalive
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Installing backend dependencies ... done
  Preparing editable metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Preparing editable metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [56 lines of output]
      INFO:hatch_jupyter_builder.utils:Running jupyter-builder
      INFO:hatch_jupyter_builder.utils:Building with hatch_jupyter_builder.npm_builder
      INFO:hatch_jupyter_builder.utils:With kwargs: {'source_dir': 'src', 'build_dir': 'jupyter_keepalive/labextension'}
      INFO:hatch_jupyter_builder.utils:Installing build dependencies with npm.  This may take a while...
      INFO:hatch_jupyter_builder.utils:> /opt/homebrew/bin/yarn install
      yarn install v1.22.21
      [1/4] Resolving packages...
      success Already up-to-date.
      $ jlpm run clean && jlpm run build:prod
      Internal Error: @minrk/jupyter-keepalive@workspace:.: This package doesn't seem to be present in your lockfile; run "yarn install" to update the lockfile
          at q0.getCandidates (/private/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/pip-build-env-h0yxotwr/overlay/lib/python3.11/site-packages/jupyterlab/staging/yarn.js:435:5145)
          at xf.getCandidates (/private/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/pip-build-env-h0yxotwr/overlay/lib/python3.11/site-packages/jupyterlab/staging/yarn.js:391:1264)
          at /private/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/pip-build-env-h0yxotwr/overlay/lib/python3.11/site-packages/jupyterlab/staging/yarn.js:439:7693
          at ff (/private/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/pip-build-env-h0yxotwr/overlay/lib/python3.11/site-packages/jupyterlab/staging/yarn.js:390:11070)
          at ge (/private/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/pip-build-env-h0yxotwr/overlay/lib/python3.11/site-packages/jupyterlab/staging/yarn.js:439:7673)
          at async Promise.allSettled (index 0)
          at async to (/private/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/pip-build-env-h0yxotwr/overlay/lib/python3.11/site-packages/jupyterlab/staging/yarn.js:390:10398)
          at async /private/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/pip-build-env-h0yxotwr/overlay/lib/python3.11/site-packages/jupyterlab/staging/yarn.js:439:8205
          at async ti.startProgressPromise (/private/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/pip-build-env-h0yxotwr/overlay/lib/python3.11/site-packages/jupyterlab/staging/yarn.js:390:47579)
          at async je.resolveEverything (/private/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/pip-build-env-h0yxotwr/overlay/lib/python3.11/site-packages/jupyterlab/staging/yarn.js:439:6233)
      error Command failed with exit code 1.
      info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
      Traceback (most recent call last):
        File "/Users/yuvipanda/.local/share/virtualenvs/jupyter-keepalive/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 167, in prepare_metadata_for_build_editable
          hook = backend.prepare_metadata_for_build_editable
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      AttributeError: module 'hatchling.build' has no attribute 'prepare_metadata_for_build_editable'
      
      During handling of the above exception, another exception occurred:
      
      Traceback (most recent call last):
        File "/Users/yuvipanda/.local/share/virtualenvs/jupyter-keepalive/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/Users/yuvipanda/.local/share/virtualenvs/jupyter-keepalive/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/Users/yuvipanda/.local/share/virtualenvs/jupyter-keepalive/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 176, in prepare_metadata_for_build_editable
          whl_basename = build_hook(metadata_directory, config_settings)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/pip-build-env-h0yxotwr/overlay/lib/python3.11/site-packages/hatchling/build.py", line 83, in build_editable
          return os.path.basename(next(builder.build(directory=wheel_directory, versions=['editable'])))
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/pip-build-env-h0yxotwr/overlay/lib/python3.11/site-packages/hatchling/builders/plugin/interface.py", line 147, in build
          build_hook.initialize(version, build_data)
        File "/private/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/pip-build-env-h0yxotwr/normal/lib/python3.11/site-packages/hatch_jupyter_builder/plugin.py", line 94, in initialize
          raise e
        File "/private/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/pip-build-env-h0yxotwr/normal/lib/python3.11/site-packages/hatch_jupyter_builder/plugin.py", line 89, in initialize
          build_func(self.target_name, version, **build_kwargs)
        File "/private/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/pip-build-env-h0yxotwr/normal/lib/python3.11/site-packages/hatch_jupyter_builder/utils.py", line 115, in npm_builder
          run([*npm_cmd, "install"], cwd=str(abs_path))
        File "/private/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/pip-build-env-h0yxotwr/normal/lib/python3.11/site-packages/hatch_jupyter_builder/utils.py", line 231, in run
          return subprocess.check_call(cmd, **kwargs)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/opt/homebrew/Cellar/python@3.11/3.11.9_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/subprocess.py", line 413, in check_call
          raise CalledProcessError(retcode, cmd)
      subprocess.CalledProcessError: Command '['/opt/homebrew/bin/yarn', 'install']' returned non-zero exit status 1.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

That turned out to be an issue with yarn versions. With correct yarn version (3), I now get:

```
Obtaining file:///Users/yuvipanda/code/jupyter-keepalive
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Installing backend dependencies ... done
  Preparing editable metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Preparing editable metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [78 lines of output]
      INFO:hatch_jupyter_builder.utils:Running jupyter-builder
      INFO:hatch_jupyter_builder.utils:Building with hatch_jupyter_builder.npm_builder
      INFO:hatch_jupyter_builder.utils:With kwargs: {'source_dir': 'src', 'build_dir': 'jupyter_keepalive/labextension'}
      INFO:hatch_jupyter_builder.utils:Installing build dependencies with npm.  This may take a while...
      INFO:hatch_jupyter_builder.utils:> /opt/homebrew/bin/yarn install
      ➤ YN0000: ┌ Resolution step
      ➤ YN0002: │ @jupyterlab/application@npm:3.6.7 doesn't provide crypto (pd8fe4), requested by @lumino/coreutils
      ➤ YN0002: │ @jupyterlab/apputils@npm:3.6.7 doesn't provide crypto (pf7a3e), requested by @lumino/coreutils
      ➤ YN0002: │ @jupyterlab/builder@npm:3.6.7 doesn't provide crypto (p8b70f), requested by @lumino/coreutils
      ➤ YN0002: │ @jupyterlab/codeeditor@npm:3.6.7 doesn't provide crypto (p44d25), requested by @lumino/coreutils
      ➤ YN0002: │ @jupyterlab/codemirror@npm:3.6.7 doesn't provide crypto (p109c1), requested by @lumino/coreutils
      ➤ YN0002: │ @jupyterlab/codemirror@npm:3.6.7 doesn't provide yjs (p56006), requested by y-codemirror
      ➤ YN0002: │ @jupyterlab/coreutils@npm:5.6.7 doesn't provide crypto (p35645), requested by @lumino/coreutils
      ➤ YN0002: │ @jupyterlab/docprovider@npm:3.6.7 doesn't provide crypto (p8dd06), requested by @lumino/coreutils
      ➤ YN0002: │ @jupyterlab/docprovider@npm:3.6.7 doesn't provide yjs (p3ed35), requested by y-protocols
      ➤ YN0002: │ @jupyterlab/docprovider@npm:3.6.7 doesn't provide yjs (p1fffb), requested by y-websocket
      ➤ YN0002: │ @jupyterlab/docregistry@npm:3.6.7 doesn't provide crypto (p3cfce), requested by @lumino/coreutils
      ➤ YN0002: │ @jupyterlab/nbformat@npm:3.6.7 doesn't provide crypto (p05ae0), requested by @lumino/coreutils
      ➤ YN0002: │ @jupyterlab/observables@npm:4.6.7 doesn't provide crypto (p3bb30), requested by @lumino/coreutils
      ➤ YN0002: │ @jupyterlab/rendermime@npm:3.6.7 doesn't provide crypto (pd8efe), requested by @lumino/coreutils
      ➤ YN0002: │ @jupyterlab/services@npm:6.6.7 doesn't provide crypto (p0dbbd), requested by @lumino/coreutils
      ➤ YN0002: │ @jupyterlab/settingregistry@npm:3.6.7 doesn't provide crypto (pfaa92), requested by @lumino/coreutils
      ➤ YN0002: │ @jupyterlab/statedb@npm:3.6.7 doesn't provide crypto (p71937), requested by @lumino/coreutils
      ➤ YN0002: │ @jupyterlab/statusbar@npm:3.6.7 doesn't provide crypto (p38e64), requested by @lumino/coreutils
      ➤ YN0002: │ @jupyterlab/translation@npm:3.6.7 doesn't provide crypto (p84294), requested by @lumino/coreutils
      ➤ YN0002: │ @jupyterlab/ui-components@npm:3.6.7 [03935] doesn't provide crypto (p963a5), requested by @lumino/coreutils
      ➤ YN0002: │ @jupyterlab/ui-components@npm:3.6.7 [8fe18] doesn't provide crypto (p4f7e5), requested by @lumino/coreutils
      ➤ YN0002: │ @lumino/application@npm:1.31.4 doesn't provide crypto (p01a46), requested by @lumino/coreutils
      ➤ YN0002: │ @lumino/commands@npm:1.21.1 doesn't provide crypto (pf7be6), requested by @lumino/coreutils
      ➤ YN0002: │ @lumino/dragdrop@npm:1.14.5 doesn't provide crypto (pe1c3c), requested by @lumino/coreutils
      ➤ YN0002: │ @lumino/polling@npm:1.11.4 doesn't provide crypto (p09645), requested by @lumino/coreutils
      ➤ YN0002: │ @lumino/widgets@npm:1.37.2 doesn't provide crypto (pa9d30), requested by @lumino/coreutils
      ➤ YN0000: │ Some peer dependencies are incorrectly met; run yarn explain peer-requirements <hash> for details, where <hash> is the six-letter p-prefixed code
      ➤ YN0000: └ Completed
      ➤ YN0000: ┌ Fetch step
      ➤ YN0000: └ Completed
      ➤ YN0000: ┌ Link step
      ➤ YN0000: └ Completed
      ➤ YN0000: Done with warnings in 0s 320ms
      INFO:hatch_jupyter_builder.utils:> /opt/homebrew/bin/yarn run build
      node_modules/@jupyterlab/rendermime/lib/widgets.d.ts(8,31): error TS2420: Class 'RenderedCommon' incorrectly implements interface 'IRenderer'.
        The types of 'title.owner.layout' are incompatible between these types.
          Type 'import("/Users/yuvipanda/code/jupyter-keepalive/node_modules/@lumino/widgets/types/layout").Layout' is not assignable to type 'import("/Users/yuvipanda/code/jupyter-keepalive/node_modules/@jupyterlab/rendermime-interfaces/node_modules/@lumino/widgets/types/layout").Layout'.
            Property '[Symbol.iterator]' is missing in type 'import("/Users/yuvipanda/code/jupyter-keepalive/node_modules/@lumino/widgets/types/layout").Layout' but required in type 'import("/Users/yuvipanda/code/jupyter-keepalive/node_modules/@jupyterlab/rendermime-interfaces/node_modules/@lumino/widgets/types/layout").Layout'.
      Traceback (most recent call last):
        File "/Users/yuvipanda/.local/share/virtualenvs/jupyter-keepalive/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 167, in prepare_metadata_for_build_editable
          hook = backend.prepare_metadata_for_build_editable
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      AttributeError: module 'hatchling.build' has no attribute 'prepare_metadata_for_build_editable'
      
      During handling of the above exception, another exception occurred:
      
      Traceback (most recent call last):
        File "/Users/yuvipanda/.local/share/virtualenvs/jupyter-keepalive/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/Users/yuvipanda/.local/share/virtualenvs/jupyter-keepalive/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/Users/yuvipanda/.local/share/virtualenvs/jupyter-keepalive/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 176, in prepare_metadata_for_build_editable
          whl_basename = build_hook(metadata_directory, config_settings)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/pip-build-env-_rh_8hwr/overlay/lib/python3.11/site-packages/hatchling/build.py", line 83, in build_editable
          return os.path.basename(next(builder.build(directory=wheel_directory, versions=['editable'])))
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/pip-build-env-_rh_8hwr/overlay/lib/python3.11/site-packages/hatchling/builders/plugin/interface.py", line 147, in build
          build_hook.initialize(version, build_data)
        File "/private/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/pip-build-env-_rh_8hwr/normal/lib/python3.11/site-packages/hatch_jupyter_builder/plugin.py", line 94, in initialize
          raise e
        File "/private/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/pip-build-env-_rh_8hwr/normal/lib/python3.11/site-packages/hatch_jupyter_builder/plugin.py", line 89, in initialize
          build_func(self.target_name, version, **build_kwargs)
        File "/private/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/pip-build-env-_rh_8hwr/normal/lib/python3.11/site-packages/hatch_jupyter_builder/utils.py", line 117, in npm_builder
          run([*npm_cmd, "run", build_cmd], cwd=str(abs_path))
        File "/private/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/pip-build-env-_rh_8hwr/normal/lib/python3.11/site-packages/hatch_jupyter_builder/utils.py", line 231, in run
          return subprocess.check_call(cmd, **kwargs)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/opt/homebrew/Cellar/python@3.11/3.11.9_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/subprocess.py", line 413, in check_call
          raise CalledProcessError(retcode, cmd)
      subprocess.CalledProcessError: Command '['/opt/homebrew/bin/yarn', 'run', 'build']' returned non-zero exit status 1.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

--- a/package.json
+++ b/package.json
@@ -45,21 +45,22 @@
     "watch:src": "tsc -w"
   },
   "dependencies": {
-    "@jupyterlab/application": "^3.1.0",
-    "@jupyterlab/coreutils": "^5.1.0",
-    "@jupyterlab/services": "^6.1.0",
-    "@jupyterlab/statusbar": "^3.5.2",
-    "@lumino/signaling": "^1.11.0",
-    "@lumino/widgets": "^1.36.0"
+    "@jupyterlab/application": "^4.1.4",
+    "@jupyterlab/coreutils": "^6.2.4",
+    "@jupyterlab/services": "^7.2.4",
+    "@jupyterlab/statusbar": "^4.2.4",
+    "@lumino/signaling": "^2.1.0",
+    "@lumino/widgets": "^2.5.0"
   },
   "devDependencies": {
-    "@jupyterlab/builder": "^3.1.0",
+    "@jupyterlab/builder": "^4.2.0",
     "eslint": "^7.14.0",
     "eslint-plugin-react": "^7.18.3",
     "mkdirp": "^1.0.3",
     "npm-run-all": "^4.1.5",
     "rimraf": "^3.0.2",
-    "typescript": "~4.1.3"
+    "typescript": "~5.0.0",
+    "yarn-deduplicate": "^6.0.2"
   },
   "jupyterlab": {
     "discovery": {

--- a/package.json
+++ b/package.json
@@ -75,5 +75,6 @@
     },
     "extension": true,
     "outputDir": "jupyter_keepalive/labextension"
-  }
+  },
+  "packageManager": "yarn@3.8.4"
 }

--- a/src/index.ts
+++ b/src/index.ts
@@ -30,7 +30,7 @@ async function keepAliveRequest(endPoint = "", init = {}) {
   let response: Response;
   try {
     response = await ServerConnection.makeRequest(requestUrl, init, settings);
-  } catch (error) {
+  } catch (error: any) {
     throw new ServerConnection.NetworkError(error);
   }
   const responseJSON = await response.text();

